Build releases from a commit hash, rather than a named branch. (#5883)
zcash/zcash: doc/release-process.md

You can optionally install the `progressbar2` Python module with pip to have a
progress bar displayed during the build process.

## Versioning

> Zcashd uses semantic version numbers as described by https://semver.org/.

Reviewer: It does not, and never has. In particular, we release breaking software changes via what SemVer would call "minor revisions", which is incompatible with SemVer. Remove this section, or change it to explain what we actually do.

`version-X.Y.0` where `X` and `Y` are the major and minor versions for the
release.

In the commands below, <RELEASE> and <RELEASE_PREV> must identify `git` tags
prefixed with the character `v`, i.e. `v1.0.9` (not `1.0.9`). <COMMIT_ID> is a
`git` hash identifying the commit on which a release stabilization or release
branch will be based. It is recommended to use the entire hash value to
identify the commit, although a prefix of at least 10 characters is also
permitted.

### Create the release stabilization branch

Having identified the commit from which the release will be made, the release
manager constructs the release stabilization branch as follows:

    $ git checkout -b version-X.Y.0 <COMMIT_ID>
    $ git push 'git@github.com:zcash/zcash' $(git rev-parse --abrev-ref HEAD)

### Create the release candidate branch

Run the release script to create the first release candidate. This will create
a branch based upon the specified commit ID, then commit standard automated
changes to that branch locally:

    $ ./zcutil/make-release.py <COMMIT_ID> <RELEASE> <RELEASE_PREV> <RELEASE_FROM> <APPROX_RELEASE_HEIGHT>

Examples:

>     $ ./zcutil/make-release.py a410f124b2 v1.0.9 v1.0.8-1 v1.0.8-1 120000
>     $ ./zcutil/make-release.py 3ee1d04a8d v1.0.13 v1.0.13-rc1 v1.0.12 222900

Reviewer: I think there's a rebase bug here; these verisons were altered in a previous PR to match what releases will now use.
```suggestion
    $ ./zcutil/make-release.py a410f124b2 v1.1.0 v1.0.0 v1.0.0 120000
    $ ./zcutil/make-release.py 3ee1d04a8d v1.1.0 v1.1.0-rc1 v1.0.0 222900
```

> ## Release process

Reviewer: Before this, we need to document the new release stabilisation process:

- Release manager creates a branch from the target commit named `version-X.Y.0` (note: without leading `v`, and no `-foo` suffix), and pushes this to `zcash/zcash`.
- Release manager immediately runs the release script on this branch, to create `release-vX.Y.0-rc1`.
  - The script could go back to verifying the branch name if we wanted, because the commit hash is still being used, just now when manually creating the stabilisation branch.
- The rc1 release branch is reviewed and merged into the stabilisation branch.
- Any stabilisation PRs are merged into the `version-X.Y.0` branch, and more RCs if needed.
- Finally, a `release-vX.Y.0` script run is created, merged, and tagged.
- At this point, the `version-X.Y.0` branch is merged back into `master` via a PR (and can be deleted, as the tags retain the necessary state).

For hotfixes:
- Release manager creates a branch from the target tag named `hotfix-vX.Y.Z`, ensuring that `X.Y` matches the target tag, and pushes this to `zcash/zcash`.
- PRs are made onto that branch as usual.
- Release manager runs the release script on this branch, to create either `release-vX.Y.Z-rc1` or `release-vX.Y.Z` (depending on whether we plan to stabilise the hotfix, or if we have already done so privately for e.g. a security fix). This gets reviewed and merged as normal.
- Same as above for next release script run, except now it's either `release-vX.Y.Z-rc2` or `release-vX.Y.Z` (and so on).

> Identify the commit from which the release will be made. This is frequently the current

Reviewer 2: ```suggestion
Identify the commit from which the release will be made. This could be the current
```

`HEAD` of `master`, but it's also often useful to instead use a release stabilization
branch based upon a previous release candidate when producing a release, so that
development can proceed unblocked on the `master` branch during the release candidate
testing and bug-fixing process.

In the commands below, <RELEASE> and <RELEASE_PREV> must be prefixed with a v, i.e. v1.0.9
(not 1.0.9). <COMMIT_ID> is the `git` hash identifying the commit on which the release
branch will be based. It is recommended to use the entire hash value to identify the

> commit, although a prefix of at least 10 characters is also permitted.

Reviewer: ```suggestion
commit, although a prefix of at least 8 characters is also permitted.
```

## Release process

Identify the commit from which the release will be made. This is frequently the current
`HEAD` of `master`, but it's also often useful to instead use a release stabilization
branch based upon a previous release candidate when producing a release, so that
development can proceed unblocked on the `master` branch during the release candidate
testing and bug-fixing process.

In the commands below, <RELEASE> and <RELEASE_PREV> must be prefixed with a v, i.e. v1.0.9
(not 1.0.9). <COMMIT_ID> is the `git` hash identifying the commit on which the release
branch will be based. It is recommended to use the entire hash value to identify the
commit, although a prefix of at least 10 characters is also permitted.

### Create the release branch

Run the release script, which will create a branch based upon the specified commit ID,
then commit standard automated changes to that branch locally:

    $ ./zcutil/make-release.py <COMMIT_ID> <RELEASE> <RELEASE_PREV> <RELEASE_FROM> <APPROX_RELEASE_HEIGHT>

Examples:

>     $ ./zcutil/make-release.py a410f12424 v1.0.9 v1.0.8-1 v1.0.8-1 120000

Reviewer 2: ```suggestion
    $ ./zcutil/make-release.py a410f124b2 v1.0.9 v1.0.8-1 v1.0.8-1 120000
```

## Post Release Task List

### Merge the release branch

Merge the release branch back to `master` to ensure that any changes made during
release stabilization are reflected in the master branch's history. 

### Deploy testnet

Notify the Zcash DevOps engineer/sysadmin that the release has been tagged. They update
some variables in the company's automation code and then run an Ansible playbook, which:

* builds Zcash based on the specified branch

> * deploys it as a public service (e.g. betatestnet.z.cash, mainnet.z.cash)

Reviewer 2: ```suggestion
* deploys it as a public service (e.g. testnet.z.cash, mainnet.z.cash)
```

* often the same server can be re-used, and the role idempotently handles upgrades, but if
  not then they also need to update DNS records
* possible manual steps: blowing away the `testnet3` dir, deleting old parameters,

>   restarting DNS seeder

Reviewer 2: ```suggestion
  restarting DNS seeder.
```

> Then, verify that nodes can connect to the testnet server, and update the guide on the
> wiki to ensure the correct hostname is listed in the recommended zcash.conf.

Reviewer 2: ```suggestion
Then, verify that nodes can connect to the testnet server, and update the
[Zcashd Full Node and CLI](https://zcash.readthedocs.io/en/latest/rtd_pages/zcashd.html)
documentation on ReadTheDocs to ensure the correct hostname is listed in
the recommended `zcash.conf`.
```

> ### Update the 1.0 User Guide

Reviewer 2: Delete this section.
